modin/core/dataframe/base/dataframe/utils.py

```diff
@@ -48,3 +52,85 @@ class JoinType(Enum):  # noqa: PR01
     LEFT = "left"
     RIGHT = "right"
     OUTER = "outer"
+
+
+def join_columns(
+    left: pandas.Index,
+    right: pandas.Index,
+    left_on: IndexLabel,
+    right_on: IndexLabel,
+    suffixes: Tuple[str, str],
+) -> Tuple[pandas.Index, Dict[IndexLabel, IndexLabel], Dict[IndexLabel, IndexLabel]]:
+    """
+    Compute resulting columns for the two dataframes being merged.
+
+    Parameters
+    ----------
+    left : pandas.Index
+        Columns of the left frame to join.
+    right : pandas.Index
+        Columns of the right frame to join.
+    left_on : list-like or scalar
+        Column names on which the frames are joined in the left DataFrame.
+    right_on : list-like or scalar
+        Column names on which the frames are joined in the right DataFrame.
+    suffixes : tuple[str, str]
+        A 2-length sequence containing suffixes to append to the intersected columns.
+
+    Returns
+    -------
+    pandas.Index, dict[IndexLabel -> IndexLabel], dict[IndexLabel -> IndexLabel]
+        Returns columns for the resulting frame and mappings of old to new column
+        names for `left` and `right` accordingly.
+
+    Raises
+    ------
+    NotImplementedError
+        Raised when one of the keys to join is an index level, pandas behaviour is really
+        complicated in this case, so we're not supporting this case for now.
+    """
+    # using `cast` to make `mypy` acknowledged that the variable now ensured to be `Sequence[IndexLabel]`
+    left_on = cast(Sequence[IndexLabel], [left_on] if is_scalar(left_on) else left_on)
+    right_on = cast(
+        Sequence[IndexLabel], [right_on] if is_scalar(right_on) else right_on
+    )
+
+    if any(col not in left for col in left_on) or any(
+        col not in right for col in right_on
+    ):
+        raise NotImplementedError(
+            "Cases, where one of the keys to join is an index level, are not yet supported."
+        )
```

Comment:
shouldn't harm HDK anyhow as it doesn't support such cases on its side
https://github.com/modin-project/modin/blob/0df5fbf7a66663110878633467e4488ad37b0337/modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py#L857-L860

Reply:
Do you mean HDK doesn't handle multi indexed columns?

Reply:
it probably handles multi-indexed columns, but it doesn't handle merging on an index level